### ISSUE TITLE
Authentication still available in Registry 2.0.0 ?

### ISSUE THREAD

**Issue by hgomez** [docker/distribution#397]
Hi all,

I'm using Registry 0.9.1 with HTTPd or NGinx in front to handle authentication.

I tried Registry 2.0.0 with docker 1.6.0 but docker client (or registry) didn't works well and client fall back to v1 

    1.2.3.203 - - [17/Apr/2015:15:41:07 +0200] "GET /v2/ HTTP/1.1" 401 1233 "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.203 - - [17/Apr/2015:15:41:07 +0200] "GET /v1/_ping HTTP/1.1" 200 1541 "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.203 - - [17/Apr/2015:15:41:07 +0200] "GET /v1/_ping HTTP/1.1" 200 1541 "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.203 - my-deployer [17/Apr/2015:15:41:07 +0200] "PUT /v1/repositories/teama1/teama-java5/ HTTP/1.1" 200 2 "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.203 - my-deployer [17/Apr/2015:15:41:07 +0200] "GET /v1/images/4f903438061c7180cf99485b42f7709f5268bfb4732fe885f9104ed3bb66fd3c/json HTTP/1.1" 200 1510 "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"

When I remove Authentication requirement on HTTPd front side, all operations works in /v2/

Nginx/Compose sample don't show authentication anymore too.

Did Registry 2.0.0 support authentication schemes ?

Thanks

**Comment by dmcgowan** [comment]
Yes registry 2.0.0 supports authentication. Do you mind commenting with the output of `docker version` as well the section of your nginx configuration that is related to the v2 registry.

**Comment by hgomez** [comment]
docker version

    Client version: 1.6.0
    Client API version: 1.18
    Go version (client): go1.4.2
    Git commit (client): 4749651
    OS/Arch (client): linux/amd64
    Server version: 1.6.0
    Server API version: 1.18
    Go version (server): go1.4.2
    Git commit (server): 4749651
    OS/Arch (server): linux/amd64

Front is using HTTPd and not nginx 

      # no proxy for /error/ (Apache HTTPd errors messages)
      ProxyPass /error/ !

      ProxyPass        /v1 http://localhost:5000/v1
      ProxyPassReverse /v1 http://localhost:5000/v1

      ProxyPass        /v2 http://localhost:5002/v2
      ProxyPassReverse /v2 http://localhost:5002/v2

      ProxyPass        /_ping http://localhost:5000/_ping
      ProxyPassReverse /_ping http://localhost:5000/_ping

      # Authentication require for push
      <Location /v1>
        Order deny,allow
        Allow from all
        AuthName "Registry Authentication"
        AuthType basic
        AuthUserFile "/etc/apache2/htpasswd/registry-htpasswd"

        # Read access to authentified users
        <Limit GET HEAD>
          Require valid-user
        </Limit>

        # Write access to ecd-deployer only
        <Limit POST PUT DELETE>
          Require user docker-deployer
        </Limit>

      </Location>

      # Allow ping to run unauthenticated.
      <Location /v1/_ping>
        Satisfy any
        Allow from all
      </Location>

      # Authentication require for push
      <Location /v2>
        Order deny,allow
        Allow from all
        AuthName "Registry Authentication"
        AuthType basic
        AuthUserFile "/etc/apache2/htpasswd/registry-htpasswd"

        # Read access to authentified users
        <Limit GET HEAD>
          Require valid-user
        </Limit>

        # Write access to ecd-deployer only
        <Limit POST PUT DELETE>
          Require user docker-deployer
        </Limit>

      </Location>

      # Allow ping to run unauthenticated.
      <Location /_ping>
        Satisfy any
        Allow from all
      </Location>

**Comment by stevvooe** [comment]
cc @moxiegirl

**Comment by hgomez** [comment]
Do you need more details, I could provides HTTPd access log (stock and forensic).

**Comment by stevvooe** [comment]
@hgomez From the logs, it's pretty clear that the v2 ping returns a 401 and falls back to v1. Are you able to hit https://registry-ancy.teama.mycorp.int/v2/ with curl or your browser using basic auth? Do you have tls configured on apache?

**Comment by hgomez** [comment]
no problem with curl 

    curl -L -v --user my-deployer:mypassword  https://registry-ancy.teama.mycorp.int/v2/ 

    * Hostname was NOT found in DNS cache
    *   Trying 1.2.3.96...
    * Connected to registry-ancy.teama.mycorp.int (1.2.3.96) port 443 (#0)
    * successfully set certificate verify locations:
    *   CAfile: none
      CApath: /etc/ssl/certs
    * SSLv3, TLS handshake, Client hello (1):
    * SSLv3, TLS handshake, Server hello (2):
    * SSLv3, TLS handshake, CERT (11):
    * SSLv3, TLS handshake, Server key exchange (12):
    * SSLv3, TLS handshake, Server finished (14):
    * SSLv3, TLS handshake, Client key exchange (16):
    * SSLv3, TLS change cipher, Client hello (1):
    * SSLv3, TLS handshake, Finished (20):
    * SSLv3, TLS change cipher, Client hello (1):
    * SSLv3, TLS handshake, Finished (20):
    * SSL connection using teamaHE-RSA-AES256-GCM-SHA384
    * Server certificate:
    * 	 subject: C=FR; ST=mycorp; L=mycorp; O=mycorp; OU=mycorp; CN=*.teama.mycorp.int; emailAddress=bczerkiewicz@mycorp.com
    * 	 start date: 2013-11-12 12:29:23 GMT
    * 	 expire date: 2015-11-12 12:29:23 GMT
    * 	 common name: *.teama.mycorp.int (matched)
    * 	 issuer: DC=int; DC=mycorp; CN=mycorpRootCA
    * 	 SSL certificate verify ok.
    * Server auth using Basic with user 'my-deployer'
    > GET /v2/ HTTP/1.1
    > Authorization: Basic ZWNkLWRlcGxveWVyOmRvY2tlcnJvY2tzaW4yMDE1
    > User-Agent: curl/7.35.0
    > Host: registry-ancy.teama.mycorp.int
    > Accept: */*
    > 
    < HTTP/1.1 200 OK
    < Date: Tue, 21 Apr 2015 08:56:44 GMT
    * Server Apache is not blacklisted
    < Server: Apache
    < Content-Length: 2
    < Content-Type: application/json; charset=utf-8
    < Docker-Distribution-Api-Version: registry/2.0
    < Host: registry-ancy.teama.mycorp.int
    < 
    * Connection #0 to host registry-ancy.teama.mycorp.int left intact

**Comment by hgomez** [comment]
Side note, 401 came back with Apache HTTPd 401 error page 

        <body>
    <h1>Authentication required!</h1>
    <p>


        This server could not verify that you are authorized to access
        the URL "/v2/".
        You either supplied the wrong credentials (e.g., bad password), or your
        browser doesn't understand how to supply the credentials required.

      </p>
    <p>

Could it be the problem ?

**Comment by hgomez** [comment]
Questions 

* Should /v2/ (ping) return 401 without credentials ?
* Do you have a working sample of v2 with httpd or nginx using httpd/nginx authentication ?

**Comment by Lispython** [comment]
I have the some problem.

Without basic auth on nginx all works well. `docker pull/push` work.

My configuration: https://gist.github.com/Lispython/36306d55bad098c2969e
Also I have chef cookbook: https://github.com/wh13/docker_registry_2
You can make VM this problem config via `kitchen converge & kitchen login`. docker-compose.yml you can find in `/var/app/`

This is my logs and tests.
https://gist.github.com/Lispython/ff10680b9e20ddd48fba

**Comment by stevvooe** [comment]
@hgomez 

> Should /v2/ (ping) return 401 when no credentials are provided ?

The /v2/ endpoint is a general auth check. It requires credentials to work correctly. There are some subtleties around fallback that need to be taken care of. @dmcgowan 

> Do you have a working sample of v2 with httpd or nginx using httpd/nginx authentication ?

Not that I know of. Let's have @moxiegirl or @dmp42 take a look.

@Lispython Do you have daemon logs, as well? It looks like the initial login is failing in the daemon and is falling back to v1.

**Comment by moxiegirl** [comment]
Sorry, I never tried httpd so I've got nothing.

**Comment by hgomez** [comment]
Did you have a working nginx configuration instead, it should be easy to adapt

**Comment by Lispython** [comment]
@stevvooe I think that docker login doesn't send Basic auth header on login.
https://gist.github.com/Lispython/ff10680b9e20ddd48fba#file-logs-output-L29 this is response from nginx, not from registry.

Please, show working nginx configuration if my is wrong.

**Comment by hgomez** [comment]
Same behaviour here, Apache HTTPd send 401, not registry.
Did there is plans to implement authentication on registry side, so we could use nginx/httpd only as relay ?

**Comment by endophage** [comment]
@hgomez There is now a PR to implement Basic auth directly in the registry using .htpasswd files https://github.com/docker/distribution/pull/409

**Comment by kuhnroyal** [comment]
I have Basic Auth working in nginx, the problem was that the docker client expects the 'Docker-Distribution-Api-Version: registry/2.0' header from the /v2/ endpoint. Setting that header in the auth response solved the problem. See example ngix config. 

```
    location /v2/ {
        # Do not allow connections from docker 1.5 and earlier
        # docker pre-1.6.0 did not properly set the user agent on ping, catch "Go *" user agents
        if ($http_user_agent ~ "^(docker\/1\.(3|4|5(?!\.[0-9]-dev))|Go ).*$" ) {
              return 404;
        }

        proxy_pass http://registry;

        proxy_read_timeout 900;

        more_set_headers 'Docker-Distribution-Api-Version: registry/2.0';
        #more_set_headers 'Content-Type: application/json; charset=utf-8';

        auth_basic "Restricted";
        auth_basic_user_file /etc/nginx/custom/dockerpasswd;
    }
```

**Comment by stevvooe** [comment]
@kuhnroyal Thank you @kuhnroyal!

@moxiegirl Could you add this to the nginx setup examples with auth?

**Comment by hgomez** [comment]
Just updated my Apache HTTPd configuration and it works too 

    Header set "Docker-Distribution-Api-Version" "registry/2.0"

all operations performed using Registry 2.0 

    1.2.3.34 - teama-deployer [22/Apr/2015:22:05:26 +0200] "GET /v2/ HTTP/1.1" 200 2 "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.34 - - [22/Apr/2015:22:05:31 +0200] "GET /v2/ HTTP/1.1" 401 1233 "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.34 - - [22/Apr/2015:22:05:32 +0200] "GET /v2/ HTTP/1.1" 401 1233 "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.34 - - [22/Apr/2015:22:05:32 +0200] "GET /v2/ HTTP/1.1" 401 1233 "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.34 - teama-deployer [22/Apr/2015:22:05:33 +0200] "HEAD /v2/teama1/teama-java5/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4 HTTP/1.1" 200 - "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.34 - teama-deployer [22/Apr/2015:22:05:33 +0200] "HEAD /v2/teama1/teama-java5/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4 HTTP/1.1" 200 - "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.34 - teama-deployer [22/Apr/2015:22:05:34 +0200] "HEAD /v2/teama1/teama-java5/blobs/sha256:b6c77cf05f323e226c196cf9c0c0011ca781d84d5005b25df8e092d5dc780ca4 HTTP/1.1" 200 - "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.34 - teama-deployer [22/Apr/2015:22:05:35 +0200] "HEAD /v2/teama1/teama-java5/blobs/sha256:cdf9bbc08a6bf9b5641961839ee9fd8beeaaacee4de55e2e360a1db364d0ba91 HTTP/1.1" 200 - "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.34 - teama-deployer [22/Apr/2015:22:05:35 +0200] "HEAD /v2/teama1/teama-java5/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4 HTTP/1.1" 200 - "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.34 - teama-deployer [22/Apr/2015:22:05:36 +0200] "HEAD /v2/teama1/teama-java5/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4 HTTP/1.1" 200 - "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.34 - teama-deployer [22/Apr/2015:22:05:36 +0200] "HEAD /v2/teama1/teama-java5/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4 HTTP/1.1" 200 - "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.34 - teama-deployer [22/Apr/2015:22:05:37 +0200] "HEAD /v2/teama1/teama-java5/blobs/sha256:153220a76b6c452aab6ce6d6874841e49f5365e2e1a242cbf37f2205db5b410a HTTP/1.1" 200 - "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.34 - teama-deployer [22/Apr/2015:22:05:37 +0200] "HEAD /v2/teama1/teama-java5/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4 HTTP/1.1" 200 - "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"
    1.2.3.34 - teama-deployer [22/Apr/2015:22:05:38 +0200] "PUT /v2/teama1/teama-java5/manifests/latest HTTP/1.1" 202 - "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.13.0-37-generic os/linux arch/amd64"

Thanks

PS: Seems to be a bug, will it be fixed in 2.0.1
REPS: If you want Apache HTTPd configuration too, just ping me :)

**Comment by stevvooe** [comment]
@hgomez Great to hear this worked!

This seems like a client bug, rather than a registry bug. I'll need to track down why we put this in exactly. I had submitted #99 but am uncertain about requiring this in the specification.

Would you be willing to contribute an example to the docs?

We'll close this when we've worked out the missing documentation issues.

cc @moxiegirl

**Comment by kuhnroyal** [comment]
@stevvooe np, that one line took me 4 hours on sunday ;)

**Comment by Lispython** [comment]
Confirm that setting header 'Docker-Distribution-Api-Version: registry/2.0' solved problem on my nginx configuration. Thanks!

**Comment by stevvooe** [comment]
@kuhnroyal @Lispython Sorry about that. I don't know about getting back your Sunday but we'll get this documented.

**Comment by moxiegirl** [comment]
@stevvooe It is on my todo list. If you need it bumped up in priority let me know. I have a few things ahead of ya.

**Comment by hgomez** [comment]
Just send a PR for HTTPd config sample

https://github.com/docker/distribution/pull/423

**Comment by andrexus** [comment]
I have the same problem. Without basic auth everything works just fine, but it stops working with basic auth enabled on the nginx side.
```
GET /v2/ HTTP/1.1" 401
```

**Comment by hgomez** [comment]
Update your nginx configuration with
    
    more_set_headers 'Docker-Distribution-Api-Version: registry/2.0'; 

It should fix problem

**Comment by hgomez** [comment]
Should we close this issue now we have a workaround ?

**Comment by andrexus** [comment]
Unfortunately your workaround didn't work. Let me explain my registry deployment.

- Registry is started (docker run -p 5000:5000 registry:2.0)
- Nginx server is started as proxy server

Here is my nginx config
```
server {
    listen 443 ssl;
    server_name registry.my.server.com;

    ssl on;
    ssl_certificate /etc/nginx/ssl/my.server_bundled.crt;
    ssl_certificate_key /etc/nginx/ssl/my.server.key;
    
    add_header Docker-Distribution-Api-Version registry/2.0;

    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Original-URI $request_uri;
    proxy_set_header Docker-Distribution-Api-Version registry/2.0;
   
    client_max_body_size 0; # disable any limits to avoid HTTP 413 for large image uploads
   
    chunked_transfer_encoding on;
   
    location / {
      auth_basic "Restricted";
      auth_basic_user_file /etc/nginx/htpasswd/users.htpasswd;
      proxy_pass http://docker-registry;
    }
}
```

I can access the path /v2/ with my username and password
```
curl  --user admin:admin https://registry.my.server.com/v2/
{}%
```
But I can't login
```
docker login -u admin -p admin -e admin@myserver.com registry.myserver.com
```
Response
```
FATA[0001] Error response from daemon: (Code: 404; Headers: map[Date:[Thu, 23 Apr 2015 14:15:32 GMT] Content-Type:[text/plain; charset=utf-8] Content-Length:[19] Connection:[keep-alive] Docker-Distribution-Api-Version:[registry/2.0] Server:[nginx/1.7.12]])
```

**Comment by hgomez** [comment]
I'm not a nginx specialist but tweak was to use 

    more_set_headers 'Docker-Distribution-Api-Version: registry/2.0';

Your configuration contains :  
    
    add_header Docker-Distribution-Api-Version registry/2.0;
    ...
    proxy_set_header Docker-Distribution-Api-Version registry/2.0;

**Comment by andrexus** [comment]
@hgomez standard nginx distribution doesn't include this module. That is why I've used
```
add_header Docker-Distribution-Api-Version registry/2.0;
...
proxy_set_header Docker-Distribution-Api-Version registry/2.0;
```
which sets request and response headers

**Comment by hgomez** [comment]
ok :)

proxy_set_header Docker-Distribution-Api-Version registry/2.0; is not enough ?

**Comment by kuhnroyal** [comment]
@hgomez As far as I understand it, proxy_set_header only will set the header if there is an actual response from the proxy target, in this case there is not, because it is being blocked by the auth request.

**Comment by hgomez** [comment]
Oups, right out and in required

**Comment by andrexus** [comment]
exactly. That means also, that registry server doesn't even get any requests because all they are blocked by nginx. The question is why does the docker client makes the requests without authorisation?

**Comment by hgomez** [comment]
Closing issue

**Comment by bluefeet** [comment]
OMG I lost probably a day to this over the last several days.  Someone on IRC (stevvooee) pointed me at this issue, and the `more_set_headers` bit did it.  Thanks all!  I normally wait for the x.0.1 versions before I dive into a major release for just these reasons.

**Comment by pkill** [comment]
Still having this issue despite the response headers being sent back to the client. I'm using a client with the same sha as @andrexus 
```
*** - - [04/May/2015:11:50:03 -0500] "GET /v2/ HTTP/1.1" 401 479 "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.19.5-100.fc20.x86_64 os/linux arch/amd64" 694
*** - - [04/May/2015:11:50:03 -0500] "GET /v1/_ping HTTP/1.1" 401 479 "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.19.5-100.fc20.x86_64 os/linux arch/amd64" 782
*** - - [04/May/2015:11:50:03 -0500] "POST /v1/users/ HTTP/1.1" 401 479 "-" "Go 1.1 package http" 713
*** - pkill [04/May/2015:11:50:03 -0500] "GET /v1/users/ HTTP/1.1" 404 19 "-" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.19.5-100.fc20.x86_64 os/linux arch/amd64" 93176
```

With or without the `Header set` directive in httpd, the response headers reflected in the error message indicate that `Docker-Distribution-Api-Version` is being properly set.

```
FATA[0006] Error response from daemon: (Code: 404; Headers: map[Date:[Mon, 04 May 2015 16:56:40 GMT] Content-Type:[text/plain; charset=utf-8] Docker-Distribution-Api-Version:[registry/2.0] Content-Length:[19] X-Content-Type-Options:[nosniff]]) 
```

I'm using httpd version 2.2.15.

**Comment by stevvooe** [comment]
@pkill Could you please ask your question on IRC, Google Groups or another issue? If you've set the header and are still having fallback problems, it might not be the same issue.

**Comment by pkill** [comment]
@stevvooe it turns out it was the same root cause: the ```Docker-Distribution-Api-Version``` isn't being set in the 401 response. I think this issue was closed prematurely and should probably be left open for until the next minor version of the registry is released. Others might run into working out the minor implementation details of this fix because of the permutations of different proxy server suites, their various configuration and versions. While the comments thus far seems to solve the problem for Nginx users.

As an httpd 2.2.15 user, I had to set two directives to get this to work.

```
    Header always set Docker-Distribution-Api-Version "registry/2.0"
    Header onsuccess set Docker-Distribution-Api-Version "registry/2.0"
``` 

According to http://httpd.apache.org/docs/2.2/mod/mod_headers.html#Header

> The optional condition argument determines which internal table of responses headers this directive will operate against. Despite the name, the default value of onsuccess...

and later

> The default value of onsuccess may need to be changed to always under the circumstances similar to those listed below. Note also that repeating this directive with both conditions makes sense in some scenarios because always is not a superset of onsuccess with respect to existing headers:

> * You're modifying or removing a header generated by some piece of the server but that header is not being found by the default onsuccess condition.

This is because httpd keeps an internal table of response headers it could send depending on a certain condition. By default, headers set with the `Header` directive will be sent when the onsuccess condition is met. However, I'm guessing that a 401 response does not count as a success condition. Setting the `always` condition places this header in a table of responses that will be sent on a 401 response. 

I will submit a pull request. 

Note to users of httpd 2.4 and later, you can refine the expression to only send the response on a 401 HTTP_STATUS condition but I have not tested that.

**Comment by hgomez** [comment]
I used HTTPd 2.4

**Comment by hgomez** [comment]
May be HTTPd 2.2 behave differently

**Comment by pkill** [comment]
@hgomez yes, that's what I'm thinking and illustrates why I think this issue should be left open until this header is no longer required.

I'm going to open a new MR to increase the compatibility of that directive. Do you have time to test my change on Apache 2.4? I'll link the MR in a bit.

**Comment by stevvooe** [comment]
@pkill I'm sorry I assumed this was a different issue. As you stated:

> the response headers reflected in the error message indicate that Docker-Distribution-Api-Version is being properly set.

This is really a documentation issue but it was closed by @hgomez, who reported it.

I'll be awaiting your PR.

**Comment by pkill** [comment]
@stevvooe No worries. The Docker client was a bit misleading when it reflected those headers in the error message because it was reflecting response headers from the /v1/ endpoint, not on the original /v2/ endpoint 401 response but as you said, it's a documentation issue.

@hgomez if you could be so kind as to test https://github.com/docker/distribution/pull/490 and comment on the issue, I would appreciate your time.

**Comment by hgomez** [comment]
I don't have apache 2.4.x anymore (moved to a new company) so cannot test it :-(

**Comment by pkill** [comment]
I have 2.4.10 on my workstation. I could probably proxy to the existing registry temporarily.

**Comment by pkill** [comment]
How odd. I can't get @hgomez's MR to work on a local instance of httpd v2.4.10. 

I didn't actually need to proxy anywhere since the initial basic auth response headers are all we care about. I also setup a host override for `127.0.0.1 registry.example.com`.

Using 
```Header set "Docker-Distribution-Api-Version" "registry/2.0"```

```
pkill@balance in {/etc/httpd/conf} at 2015-05-04 16:30:25
$> curl -v -s "http://registry.example.com/v2/"  1> /dev/null
* Adding handle: conn: 0x96e8d0
* Adding handle: send: 0
* Adding handle: recv: 0
* Curl_addHandleToPipeline: length: 1
* - Conn 0 (0x96e8d0) send_pipe: 1, recv_pipe: 0
* About to connect() to registry.example.com port 80 (#0)
*   Trying 127.0.0.1...
* Connected to registry.example.com (127.0.0.1) port 80 (#0)
> GET /v2/ HTTP/1.1
> User-Agent: curl/7.32.0
> Host: registry.example.com
> Accept: */*
> 
< HTTP/1.1 401 Unauthorized
< Date: Mon, 04 May 2015 21:30:33 GMT
* Server Apache/2.4.10 (Fedora) OpenSSL/1.0.1e-fips is not blacklisted
< Server: Apache/2.4.10 (Fedora) OpenSSL/1.0.1e-fips
< WWW-Authenticate: Basic realm="Registry Authentication"
< Content-Length: 381
< Content-Type: text/html; charset=iso-8859-1
< 
{ [data not shown]
* Connection #0 to host registry.example.com left intact
```

Using
```
Header always set "Docker-Distribution-Api-Version" "registry/2.0"
Header onsuccess set "Docker-Distribution-Api-Version" "registry/2.0"
```

```
pkill@balance in {/etc/httpd/conf} at 2015-05-04 16:32:08
$> sudo systemctl restart httpd                                                                                                                                            

pkill@balance in {/etc/httpd/conf} at 2015-05-04 16:32:15
$> curl -v -s "http://registry.example.com/v2/"  1> /dev/null
* Adding handle: conn: 0x13e58d0
* Adding handle: send: 0
* Adding handle: recv: 0
* Curl_addHandleToPipeline: length: 1
* - Conn 0 (0x13e58d0) send_pipe: 1, recv_pipe: 0
* About to connect() to registry.example.com port 80 (#0)
*   Trying 127.0.0.1...
* Connected to registry.example.com (127.0.0.1) port 80 (#0)
> GET /v2/ HTTP/1.1
> User-Agent: curl/7.32.0
> Host: registry.example.com
> Accept: */*
> 
< HTTP/1.1 401 Unauthorized
< Date: Mon, 04 May 2015 21:32:16 GMT
* Server Apache/2.4.10 (Fedora) OpenSSL/1.0.1e-fips is not blacklisted
< Server: Apache/2.4.10 (Fedora) OpenSSL/1.0.1e-fips
< Docker-Distribution-Api-Version: registry/2.0
< WWW-Authenticate: Basic realm="Registry Authentication"
< Content-Length: 381
< Content-Type: text/html; charset=iso-8859-1
< 
{ [data not shown]
* Connection #0 to host registry.example.com left intact
```

¯\\\_(ツ)\_/¯

**Comment by tnolet** [comment]
This info should be right there in the Docker Registry 2.0 documentation. Quite a ride to get this working with basic authentication and SSL. Maybe even pre-package a container that houses these components together. I know it's against the Docker ethos, but you kinda always need a proxy besides some locat test as far as I can see.

**Comment by stevvooe** [comment]
@tnolet Where exactly in the documentation should such a header requirement be placed? The registry sets it appropriately, but the proxies seem to handle header configuration differently in each case. To put it another way, where would you have found it, had it been where it was supposed to be?

@dave-tucker @moxiegirl

**Comment by tnolet** [comment]
@stevvooe just a little extra mention in the install docs at https://docs.docker.com/registry/deploying/ would suffice. Maybe put it in one of those alert/info boxes. Maybe add the part for basic authentication to the same page would really help.

**Comment by hgomez** [comment]
Docker 1.6.1 and Distribution 2.0.1 have been released and I'm wondering if nginx/httpd tricks are still required ?

If so, which version of Docker and Distribution should fix header issues ?

Thanks

**Comment by dmcgowan** [comment]
@hgomez don't expect Docker 1.6.x to be able to connect to a v2 registry without the  "Docker-Distribution-Api-Version".  We are trying to remove this requirement for 1.7 but we need to be able to get the new registry discovery process into the engine (as defined here #303). We have no solution which we can define on the distribution server which would mitigate the need for this header.

**Comment by gauravbhandare** [comment]
This issue seems to be reintroduced in docker 1.8.2 onwards. With docker registry V2 fronted by nginx, it works fine with clients <= 1.7 but with 1.8.2 it gives the error. Also tried the same with 1.9 rc but issue persists.

**Comment by stevvooe** [comment]
@gauravbhandare This issue has been long closed. Generally, commenting on closed issues is considered bad etiquette unless you're 100% sure that the issue is the same. In this case, you are describing an issue that has similar behavior but the root cause is likely very different.

Please open a new issue or consult mailing list to get help with your issue. If you believe this issue is related, you should reference it in your new issue and we can compare the symptoms.
